### ISSUE TITLE
Changing ARCHLVL does not do sysclear or sysreset clear

### ISSUE THREAD

**Issue by Fish-Git** [Fish-Git/hyperion#92]
I have had problems in the past when trying to develop 370 runtests which do I/O and then wait for the interrupt by loading an enabled wait PSW. The interrupt would never occur and the test would fail with a timeout.

An `ipending` display showed the CPU _was_ enabled for I/O interrupts, and the interrupt _was_ queued.

It was only when I displayed my control registers that I realized what was going on: Control Register 2 (Channel masks) was _all zeroes!_ (wtf?)

It appears switching architecture modes (`archlvl` command), for some reason, does **_not_** do an automatic `sysclear` (or `sysreset clear`). It was only when I added a `sysclear` statement immediately _after_ my `archlvl` statement that things began working properly due to the sysclear resetting CR2 to all ones the way it should have been initially.

I am _strongly_ considering adding code to do an automatic `sysclear` function whenever the architecture mode is changed. This would include the automatic switch to ESA/390 mode during a z/Arch IPL as long as doing so wouldn't mess anything up (which I can't see how it would).

My justification for such a change is that the `archlvl` command/statement is supposed to define the **_initial_** architecture mode (i.e. the mode that the system should be in immediately upon power-on impl). Once the _initial_ architecture mode is set, then from then on it channot be changed except automatically as a result of the automatic switch to ESA/390 mode that occurs during a z/Architecture IPL or via a `SIGP` (Signal Processor) instruction.

Does anyone see a problem with myproposed plan to perform an automatic `sysclear` function whenever the `archlvl` is changed?

**Comment by Fish-Git** [comment]
Here is a Hercules log that illustrates the problem:

```
15:38:13 HHC01603I archlvl
15:38:13 HHC02203I ARCHLVL       : z/Arch
15:38:18 HHC01603I cr
15:38:18 HHC02271I Control registers
15:38:18 HHC02271I C0=0000000000000060 C1=0000000000000000 C2=0000000000000000 C3=0000000000000000
15:38:18 HHC02271I C4=0000000000000000 C5=0000000000000000 C6=0000000000000000 C7=0000000000000000
15:38:18 HHC02271I C8=0000000000000000 C9=0000000000000000 CA=0000000000000000 CB=0000000000000000
15:38:18 HHC02271I CC=0000000000000000 CD=0000000000000000 CE=00000000C2000000 CF=0000000000000000
15:38:32 HHC01603I archlvl 370
15:38:32 HHC00811I Processor CP00: architecture mode S/370
15:38:32 HHC02204I ARCHLVL        set to S/370
15:38:32 HHC02204I LPARNUM        set to BASIC
15:38:37 HHC01603I cr
15:38:37 HHC02271I Control registers
15:38:37 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:38:37 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:38:37 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:38:37 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:38:48 HHC01603I sysreset
15:38:48 HHC02311I sysreset completed
15:38:50 HHC01603I cr
15:38:50 HHC02271I Control registers
15:38:50 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:38:50 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:38:50 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:38:50 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:39:00 HHC01603I sysreset clear
15:39:00 HHC02311I sysreset clear completed
15:39:03 HHC01603I cr
15:39:03 HHC02271I Control registers
15:39:03 HHC02271I CR00=000000E0 CR01=00000000 CR02=FFFFFFFF CR03=00000000
15:39:03 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:39:03 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:39:03 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000200
15:39:13 HHC01603I archlvl z/arch
15:39:13 HHC00811I Processor CP00: architecture mode z/Arch
15:39:13 HHC02204I ARCHLVL        set to z/Arch
15:39:13 HHC02204I LPARNUM        set to 1
15:39:13 HHC02204I CPUIDFMT       set to 0
15:39:17 HHC01603I cr
15:39:17 HHC02271I Control registers
15:39:17 HHC02271I C0=00000000000000E0 C1=0000000000000000 C2=00000000FFFFFFFF C3=0000000000000000
15:39:17 HHC02271I C4=0000000000000000 C5=0000000000000000 C6=0000000000000000 C7=0000000000000000
15:39:17 HHC02271I C8=0000000000000000 C9=0000000000000000 CA=0000000000000000 CB=0000000000000000
15:39:17 HHC02271I CC=0000000000000000 CD=0000000000000000 CE=00000000C2000000 CF=0000000000000200
15:39:23 HHC01603I sysclear
15:39:23 HHC00811I Processor CP00: architecture mode ESA/390
15:39:23 HHC02311I sysclear completed
15:39:25 HHC01603I cr
15:39:25 HHC02271I Control registers
15:39:25 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:39:25 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:39:25 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:39:25 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:39:39 HHC01603I sysclear
15:39:39 HHC02311I sysclear completed
15:39:41 HHC01603I cr
15:39:41 HHC02271I Control registers
15:39:41 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:39:41 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:39:41 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:39:41 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:39:57 HHC01603I archlvl z/arch
15:39:57 HHC00811I Processor CP00: architecture mode z/Arch
15:39:57 HHC02204I ARCHLVL        set to z/Arch
15:40:02 HHC01603I sysclear
15:40:02 HHC00811I Processor CP00: architecture mode ESA/390
15:40:02 HHC02311I sysclear completed
15:40:05 HHC01603I cr
15:40:05 HHC02271I Control registers
15:40:05 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:40:05 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:40:05 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:40:05 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:40:19 HHC01603I archlvl z/arch
15:40:19 HHC00811I Processor CP00: architecture mode z/Arch
15:40:19 HHC02204I ARCHLVL        set to z/Arch
15:40:21 HHC01603I cr
15:40:21 HHC02271I Control registers
15:40:21 HHC02271I C0=0000000000000060 C1=0000000000000000 C2=0000000000000000 C3=0000000000000000
15:40:21 HHC02271I C4=0000000000000000 C5=0000000000000000 C6=0000000000000000 C7=0000000000000000
15:40:21 HHC02271I C8=0000000000000000 C9=0000000000000000 CA=0000000000000000 CB=0000000000000000
15:40:21 HHC02271I CC=0000000000000000 CD=0000000000000000 CE=00000000C2000000 CF=0000000000000000
15:40:43 HHC01603I sysreset
15:40:43 HHC02311I sysreset completed
15:40:45 HHC01603I cr
15:40:45 HHC02271I Control registers
15:40:45 HHC02271I C0=0000000000000060 C1=0000000000000000 C2=0000000000000000 C3=0000000000000000
15:40:45 HHC02271I C4=0000000000000000 C5=0000000000000000 C6=0000000000000000 C7=0000000000000000
15:40:45 HHC02271I C8=0000000000000000 C9=0000000000000000 CA=0000000000000000 CB=0000000000000000
15:40:45 HHC02271I CC=0000000000000000 CD=0000000000000000 CE=00000000C2000000 CF=0000000000000000
15:40:58 HHC01603I archlvl 370
15:40:58 HHC00811I Processor CP00: architecture mode S/370
15:40:58 HHC02204I ARCHLVL        set to S/370
15:40:58 HHC02204I LPARNUM        set to BASIC
15:41:00 HHC01603I cr
15:41:00 HHC02271I Control registers
15:41:00 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:41:00 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:41:00 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:41:00 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:41:06 HHC01603I sysreset
15:41:06 HHC02311I sysreset completed
15:41:08 HHC01603I cr
15:41:08 HHC02271I Control registers
15:41:08 HHC02271I CR00=00000060 CR01=00000000 CR02=00000000 CR03=00000000
15:41:08 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:41:08 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:41:08 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000000
15:41:15 HHC01603I sysreset clear
15:41:15 HHC02311I sysreset clear completed
15:41:17 HHC01603I cr
15:41:17 HHC02271I Control registers
15:41:17 HHC02271I CR00=000000E0 CR01=00000000 CR02=FFFFFFFF CR03=00000000
15:41:17 HHC02271I CR04=00000000 CR05=00000000 CR06=00000000 CR07=00000000
15:41:17 HHC02271I CR08=00000000 CR09=00000000 CR10=00000000 CR11=00000000
15:41:17 HHC02271I CR12=00000000 CR13=00000000 CR14=C2000000 CR15=00000200
```

As you can see, S/370's CR0 is _wrong_ immediately after switching to `archlvl 370` until the `sysclear` is eventually done. Only then is it correct. This is the specific problem I am trying to fix, but I believe the fix really should apply to any/all architecture mode switches and not to just S/370. Yes?

**Comment by Fish-Git** [comment]
(initial and followup posting corrected to specify CR2, not CR0! Typo. Sorry!)

**Comment by Fish-Git** [comment]
> Once the initial architecture mode is set, then from then on it channot be changed except automatically as a result of the automatic switch to ESA/390 mode that occurs during a z/Architecture IPL or via a SIGP (Signal Processor) instruction.

Except by another `archlvl` command of course! (which cannot be done (is rejected) one the guest is IPLed).

**Comment by Fish-Git** [comment]
Resolved by commits 4298fc94f92d382e505441f26b5215da38663a02 and e5c14b1870a67e6002377103d4814e750602cda0: the `ARCHLVL` command processing function was corrected to ensure an initial CPU reset is always performed on _all_ CPUs whenever _any_ architecture is set (i.e. _regardless_ of whether the architecture being set is the same as, or is different from, the currently defined architecture).
